File: mifosng-android/src/main/java/com/mifos/mifosxdroid/online/loanaccountsummary/LoanAccountSummaryPresenter.java
package com.mifos.mifosxdroid.online.loanaccountsummary;

import com.mifos.api.datamanager.DataManagerLoan;
import com.mifos.mifosxdroid.base.BasePresenter;
import com.mifos.objects.accounts.loan.LoanWithAssociations;

import javax.inject.Inject;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.observers.DisposableObserver;
import io.reactivex.schedulers.Schedulers;


/**
 * Created by Rajan Maurya on 07/06/16.
 */
public class LoanAccountSummaryPresenter extends BasePresenter<LoanAccountSummaryMvpView> {

    private final DataManagerLoan mDataManagerLoan;
    private CompositeDisposable compositeDisposable;

    @Inject
    public LoanAccountSummaryPresenter(DataManagerLoan dataManagerLoan) {
        mDataManagerLoan = dataManagerLoan;
        compositeDisposable = new CompositeDisposable();
    }

    @Override
    public void attachView(LoanAccountSummaryMvpView mvpView) {
        super.attachView(mvpView);
    }

    @Override
    public void detachView() {
        super.detachView();
        compositeDisposable.clear();
    }

    public void loadLoanById(int loanAccountNumber) {
        checkViewAttached();
        getMvpView().showProgressbar(true);
        compositeDisposable.add(mDataManagerLoan.getLoanById(loanAccountNumber)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribeWith(new DisposableObserver<LoanWithAssociations>() {
                    @Override
                    public void onComplete() {
                    }

                    @Override
                    public void onError(Throwable e) {
                        getMvpView().showProgressbar(false);
                        getMvpView().showFetchingError("Loan Account not found.");
                    }

                    @Override
                    public void onNext(LoanWithAssociations loanWithAssociations) {
                        getMvpView().showProgressbar(false);
                        getMvpView().showLoanById(loanWithAssociations);
                    }
                }));
    }

}

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/online/loanaccountsummary/LoanAccountSummaryPresenter.kt
package com.mifos.mifosxdroid.online.loanaccountsummary

import com.mifos.App
import com.mifos.api.datamanager.DataManagerLoan
import com.mifos.mifosxdroid.R
import com.mifos.mifosxdroid.base.BasePresenter
import com.mifos.objects.accounts.loan.LoanWithAssociations
import com.mifos.utils.Network
import rx.Subscriber
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import rx.subscriptions.CompositeSubscription
import javax.inject.Inject

/**
 * Created by Rajan Maurya on 07/06/16.
 */
class LoanAccountSummaryPresenter @Inject constructor(private val mDataManagerLoan: DataManagerLoan) : BasePresenter<LoanAccountSummaryMvpView?>() {
    private val mSubscriptions: CompositeSubscription
    override fun attachView(mvpView: LoanAccountSummaryMvpView?) {
        super.attachView(mvpView)
    }

    override fun detachView() {
        super.detachView()
        mSubscriptions.unsubscribe()
    }

    fun loadLoanById(loanAccountNumber: Int) {
        checkViewAttached()
        mvpView!!.showProgressbar(true)
        mSubscriptions.add(mDataManagerLoan.getLoanById(loanAccountNumber)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(object : Subscriber<LoanWithAssociations?>() {
                    override fun onCompleted() {}
                    override fun onError(e: Throwable) {
                        var error = App.getContext().getString(R.string.loan_acc_not_found)
                        mvpView!!.showProgressbar(false)
                        if (!Network.isOnline(App.getContext())) {
                            error = App.getContext().getString(R.string.no_internet_connection)
                        }
                        mvpView!!.showProgressbar(false)
                        mvpView!!.showFetchingError(error)
                    }

                    override fun onNext(loanWithAssociations: LoanWithAssociations?) {
                        mvpView!!.showProgressbar(false)
                        loanWithAssociations?.let { mvpView!!.showLoanById(it) }
                    }
                }))
    }

    init {
        mSubscriptions = CompositeSubscription()
    }
}